cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/DefaultPermissionManagement.java

```diff
@@ -120,15 +122,72 @@ public PermissionCheckResult getPermissionResult(@NotNull IPermissible permissib
     @Override
     @NotNull
     public PermissionCheckResult getPermissionResult(@NotNull IPermissible permissible, @NotNull Permission permission) {
-        return this.getPermissionResult(permissible, () -> permissible.hasPermission(permission), permissionGroup -> this.tryExtendedGroups(permissionGroup.getName(), permissionGroup, permission, 0));
+        Permission bestMatch = permissible.findMatchingPermission(permissible.getPermissions(), permission);
+        Permission bestGroupMatch = tryExtendedGroups(bestMatch == null ? permission : bestMatch, permissible, null, new HashSet<>());
+
+        if (bestMatch == null || bestGroupMatch == null) {
+            return PermissionCheckResult.fromPermission(bestMatch == null ? bestGroupMatch : bestMatch);
+        } else {
+            return PermissionCheckResult.fromPermission(bestGroupMatch.compareTo(bestMatch) > 0 ? bestGroupMatch : bestMatch);
+        }
     }
 
     @Override
     @NotNull
     public PermissionCheckResult getPermissionResult(@NotNull IPermissible permissible, @NotNull String group, @NotNull Permission permission) {
-        return this.getPermissionResult(permissible, () -> permissible.hasPermission(permission), permissionGroup -> this.tryExtendedGroups(permissionGroup.getName(), permissionGroup, group, permission, 0));
+        Collection<Permission> permissions = permissible.getGroupPermissions().get(group);
+        if (permissions != null) {
+            Permission bestMatch = permissible.findMatchingPermission(permissions, permission);
+            Permission bestGroupMatch = tryExtendedGroups(bestMatch == null ? permission : bestMatch, permissible, group, new HashSet<>());
+
+            if (bestMatch == null || bestGroupMatch == null) {
+                return PermissionCheckResult.fromPermission(bestMatch == null ? bestGroupMatch : bestMatch);
+            } else {
+                return PermissionCheckResult.fromPermission(bestGroupMatch.compareTo(bestMatch) > 0 ? bestGroupMatch : bestMatch);
+            }
+        }
+        return PermissionCheckResult.DENIED;
     }
 
+    private Permission tryExtendedGroups(Permission permission, IPermissible permissible, String currentGroup, Set<String> testedGroups) {
+        Permission bestResult = null;
+        for (IPermissionGroup group : this.getGroups(permissible)) {
```

Comment:
Replace wirh a while loop and iterator based on the best result boolean. To reduce for/while-loop operations